provider/cmd/crd2pulumi/gen/dotnet.go

```diff
@@ -0,0 +1,203 @@
+		group, version := splitGroupVersion(groupVersion)
+		groupPrefix := groupPrefix(group)
+		namespaces[groupVersion] = strings.Title(groupPrefix) + "." + versionToUpper(version)
+	}
+	namespaces["meta/v1"] = "Meta.V1"
+
+	// Configure C# language-specific settings. Notice that we set
+	// `compatibility` to `kubernetes20`. This is because the actual ObjectMeta
+	// class defined in the .NET SDK is located at
+	// `Pulumi.Kubernetes.Types.Outputs.Meta.V1.ObjectMeta`. This path would
+	// only get generated properly if `compatibility` was `kubernetes20`.
+	pkg.Language["csharp"] = rawMessage(map[string]interface{}{
+		"compatibility":          "kubernetes20",
+		"dictionaryConstructors": true,
+		"namespaces":             namespaces,
+	})
+
+	files, err := dotnet.GeneratePackage(tool, pkg, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not generate .NET package")
+	}
+
+	files["KubernetesResource.cs"] = []byte(kubernetesResource)
+	files["Utilities.cs"] = []byte(dotNetUtilities)
+
+	// Delete unneeded files
+	for _, unneededFile := range unneededDotNetFiles {
+		delete(files, unneededFile)
+	}
+
+	buffers := map[string]*bytes.Buffer{}
+	for name, code := range files {
+		buffers[name] = bytes.NewBuffer(code)
+	}
+
+	return buffers, nil
+}
+
+// Returns a set of the file paths for each CustomResource class for a generated
+// C# package.
+func (pg *PackageGenerator) dotNetClassPaths() codegen.StringSet {
+	paths := codegen.NewStringSet()
+	for _, crg := range pg.CustomResourceGenerators {
+		for _, groupVersion := range crg.GroupVersions {
+			group, version := splitGroupVersion(groupVersion)
+			path := filepath.Join(strings.Title(groupPrefix(group)), versionToUpper(version), crg.Kind+".cs")
+			paths.Add(path)
+		}
+	}
+	return paths
+}
+
+const kubernetesResource = `// Copyright 2016-2020, Pulumi Corporation
+
+namespace Pulumi.Kubernetes
+{
+    /// <summary>
+    /// A base class for all Kubernetes resources.
+    /// </summary>
+    public abstract class KubernetesResource : CustomResource
+    {
+        /// <summary>
+        /// Standard constructor passing arguments to <see cref="CustomResource"/>.
+        /// </summary>
+        internal KubernetesResource(string type, string name, ResourceArgs? args, CustomResourceOptions? options = null)
+            : base(type, name, args, options)
+        {
+        }
+
+        /// <summary>
+        /// Additional constructor for dynamic arguments received from YAML-based sources.
+        /// </summary>
+        internal KubernetesResource(string type, string name, DictionaryResourceArgs? args, CustomResourceOptions? options = null)
+            : base(type, name, args, options)
+        {
+        }
+    }
+}
+`
+
+// For some reason, we get a `Missing embedded version.txt file` error if we
+// tried running `pulumi up` with the normal `Utilities.cs` file.
+// As a temporary fix, this modified `Utilities.cs` file just removes the
+// `static Utilities()` method.
```

Comment:
I think we're going to need to come up with a better solution, because creating the resource sets the `CustomResourceOptions.Version` to the value returned from `Utilities.Version`.

```c#
        public CronTab(string name, Pulumi.Kubernetes.Types.Inputs.Stable.V1.CronTabArgs? args = null, CustomResourceOptions? options = null)
            : base("kubernetes:stable.example.com/v1:CronTab", name, MakeArgs(args), MakeResourceOptions(options, ""))
        {
        }

        private static CustomResourceOptions MakeResourceOptions(CustomResourceOptions? options, Input<string>? id)
        {
            var defaultOptions = new CustomResourceOptions
            {
                Version = Utilities.Version,
            };
            var merged = CustomResourceOptions.Merge(defaultOptions, options);
            // Override the ID if one was specified for consistency with other language SDKs.
            merged.Id = id ?? merged.Id;
            return merged;
        }
```

We might need to update the implementation of Utilities that we emit here to have `Utilities.Version` call the `Utilities.Version` inside the Kubernetes assembly using private reflection. Alternatively, we may need to add a new flag to pass to the codegen in `pulumi/pulumi` for .NET to not set the `CustomResourceOptions.Version` at all in any emitted resources, if it's OK not to set it (not sure why setting it is necessary). Curious if @mikhailshilkov has an opinion here.

I think we may also be able to pare down this copy of the class -- I don't think `GetEnv*` and `WithVersion` are needed here.

Reply:
> Kubernetes assembly using private reflection

Assembly version isn't always equal to the provider version, unfortunately. I know they are different for alpha-generated versions, when a provider version isn't a legal assembly version.

Anyway, it looks like we need another flag for .NET codegen of CRDs:
- Don't generate `Utilities`, `MakeResourceOptions`, etc.
- Don't use `KubernetesResource` base class
- Maybe, use the standard namespacing approach but be aware of k8s-specific namespacing for existing classes like metadata
- Rename `csproj`
- ...

Ultimately, custom resources are different from our own resources, and it looks like we can't ignore that.

Was this not necessary for other languages?

Reply:
Although TypeScript used the generated `Utilities` file, it never ran into an error somehow. For Go, it seemed like the Utilities file was never even referenced by the CustomResources in the first place. For Python I got an error with Utilities, but since the Utilities file just contained functions, I was able to import the same utilities file from the actual SDK and have the generated utilities file just call those same functions instead (not sure if this is an actual solution though).